synapse/http/client.py

```diff
@@ -732,35 +764,27 @@ def connectionLost(self, reason):
             self.deferred.errback(reason)
 
 
-# XXX: FIXME: This is horribly copy-pasted from matrixfederationclient.
-# The two should be factored out.
-
-
-def _readBodyToFile(response, stream, max_size):
+def _readBodyToFile(
+    response: IResponse, stream: BinaryIO, max_size: Optional[int]
+) -> defer.Deferred:
     d = defer.Deferred()
     response.deliverBody(_ReadBodyToFileProtocol(stream, d, max_size))
     return d
 
 
-def encode_urlencode_args(args):
-    return {k: encode_urlencode_arg(v) for k, v in args.items()}
-
+def encode_query_args(args: Optional[Mapping[str, Union[str, List[str]]]]) -> bytes:
+    if args is None:
+        return b""
 
-def encode_urlencode_arg(arg):
-    if isinstance(arg, str):
-        return arg.encode("utf-8")
-    elif isinstance(arg, list):
-        return [encode_urlencode_arg(i) for i in arg]
-    else:
-        return arg
```

Comment:
Can we be absolutely certain that nothing was relying on being able to pass in `bytes`es here?

Reply:
This was only used by `post_urlencoded_get_json`, which is called in two places:

* [`IdentityHandler.ask_id_server_for_third_party_invite`](https://github.com/matrix-org/synapse/blob/1c262431f9bf768d106bf79a568479fa5a0784a1/synapse/handlers/identity.py#L858-L860)
    * Following this up the chain of callers it seems the keys and values are always strings.
* [`RecaptchaAuthChecker.check_auth`](https://github.com/matrix-org/synapse/blob/c9c0ad5e204f309f2686dbe250382e481e0f82c2/synapse/handlers/ui_auth/checkers.py#L108)
    * Following the callers it seems like the data is strings, in particular it seems the [response is a string and not bytes](https://github.com/matrix-org/synapse/blob/514a240aedd972e6c8b707a7685c3d283f8618c9/synapse/rest/client/v2_alpha/auth.py#L129-L134)

Would appreciate a double checking of this though.

Reply:
yeah, seems fair to me :+1: